src/libopenrave/kinbodystatesaver.cpp

```diff
@@ -18,6 +18,121 @@
+        RAVELOG_WARN_FORMAT("env=%s, could not restore link '%s' since its index %d is out of range (body num links is %d)", pEnv->GetNameId()%pLinkRef->GetName()%pLinkRef->GetIndex()%vLinks.size());
+    }
+    else {
+        grabbed._listNonCollidingLinksWhenGrabbed.push_back(vLinks.at(linkindex));
+    }
+}
+
+void KinBody::_RestoreGrabbedBodiesFromSavedData(const KinBody& savedBody,
+                                                 const int options,
+                                                 const std::unordered_map<int, GrabbedPtr>& savedGrabbedBodiesByEnvironmentIndex)
+{
+    {
+        // have to release all grabbed first
+        ReleaseAllGrabbed();
+        OPENRAVE_ASSERT_OP(_grabbedBodiesByEnvironmentIndex.size(),==,0);
+        for (const std::unordered_map<int, GrabbedPtr>::value_type& grabPair : savedGrabbedBodiesByEnvironmentIndex) {
+            const GrabbedPtr& pGrabbed = grabPair.second;
+            KinBodyPtr pGrabbedBody = pGrabbed->_pGrabbedBody.lock();
+            if( !!pGrabbedBody ) {
+                KinBody::LinkPtr pGrabbingLink(pGrabbed->_pGrabbingLink);
+                if( !pGrabbingLink ) {
+                    throw OPENRAVE_EXCEPTION_FORMAT(_("env=%s, could not find grabbing link for grabbed body '%s'"),
+                                                    GetEnv()->GetNameId()%pGrabbedBody->GetName(),
+                                                    ORE_Failed);
+                }
+                else if( !GetLink(pGrabbingLink->GetName()) ) {
+                    throw OPENRAVE_EXCEPTION_FORMAT(_("env=%s, could not find grabbing link '%s' for grabbed body '%s' since %s does not have the link (body num links is %d)"),
+                                                    GetEnv()->GetNameId()%pGrabbingLink->GetName()%pGrabbedBody->GetName()%GetName()%GetLinks().size(),
+                                                    ORE_Failed);
+                }
+                if( GetEnv() == savedBody.GetEnv() ) {
+                    // attach and add
+                    _AttachBody(pGrabbedBody);
+                    BOOST_ASSERT(pGrabbedBody->GetEnvironmentBodyIndex() > 0);
+                    _grabbedBodiesByEnvironmentIndex[pGrabbedBody->GetEnvironmentBodyIndex()] = pGrabbed;
+
+                    // grabbed bodies could have been removed from env and self collision checker.
+                    CollisionCheckerBasePtr collisionchecker = GetSelfCollisionChecker();
+                    if (!!collisionchecker) {
+                        collisionchecker->InitKinBody(pGrabbedBody);
+                    }
+                }
+                else {
+                    // The body that the state was saved from is from a different environment from savedBody. This case can
+                    // happen when cloning an environment (see Environment::_Clone). Since cloning is supposed to
+                    // preserve environmentBodyIndex of bodies, it is ok do the following.
```

Comment:
Is it possible to clone an env that has state savers that reference body N, remove body N from the original environment, add a new body that then gets reassigned that index, and _then_ have the state saver trigger? What happens in this case? 
Actually, why do the cloned bodies / state savers point to the original environment? Wouldn't it be better if they all pointed to bodies within their own env?

Reply:
Thanks for your comments!

Actually, your concerns are very good points.

> Is it possible to clone an env that has state savers that reference body N, remove body N from the original environment, add a new body that then gets reassigned that index, and then have the state saver trigger? What happens in this case?

It's kind of undefined behavior.

> Actually, why do the cloned bodies / state savers point to the original environment? Wouldn't it be better if they all pointed to bodies within their own env?

Yea, totally agree.


This code is already in `production` branch mostly.
It's purely used from `Environment::_Clone`.
https://github.com/rdiankov/openrave/blob/production/src/libopenrave-core/environment-core.h#L3883-L3884

As for the former way, the `_Clone` function seems carefully written to preserve the `envBodyIndex` of the original env.
So, this use case is barely ok.

But, if the user accidentally call the saver's functionality, it's not safe in general.
And, unfortunately, there seems no way to detect such wrong use case inside of saver.

Thus, in another PR, I deprecate the saver to restore the grabbed bodies from one env to another env. I still keep the use case from `_Clone`:
https://github.com/rdiankov/openrave/pull/1431